File: MatlabOctaveCodeSnippets/snip_frequency_pwelch.m
N = 3000; %total number of signal samples
Nfft = 32; %segment length M and also FFT-length
Fs = 1; %assumed bandwidth for discrete-time signal
power_x = 600; %noise power in Watts
x=sqrt(power_x) * randn(1,N); %Gaussian white noise
[Sk,F]=pwelch(x,hamming(Nfft),[],Nfft,Fs,'twosided'); %Parameter []
%is because in Matlab it is num. samples while Octave is percent (%)
disp(['Periodogram standard deviation=' num2str(std(Sk))])
plot(2*pi*F,Sk) %periodogram with Fs=1 is discrete-time PSD
hold on, plot(w,power_x*ones(1,length(w)),'r') %PSD theoretical value

File: PythonCodeSnippets/snip_frequency_pwelch.py
import numpy as np
from scipy.signal import welch, windows
import matplotlib.pyplot as plt

N = 3000  # total number of signal samples
Nfft = 32  # segment length M and also FFT-length
Fs = 1  # assumed bandwidth for discrete-time signal
power_x = 600  # noise power in Watts

# Gaussian white noise
x = np.sqrt(power_x) * np.random.randn(N)

# Parameter [] is because in Matlab it is num. samples while Octave is percent (%)
f, Sk = welch(x, window=windows.hamming(Nfft), nperseg=Nfft, fs=Fs)

print("Periodogram standard deviation=", np.std(Sk))

# periodogram with Fs=1 is discrete-time PSD
plt.plot(2 * np.pi * f, Sk)

# PSD theoretical value
plt.plot(f, power_x * np.ones_like(f), "r")
plt.show()
